### ISSUE TITLE
Weight_decay in SGD doesn't respect hooks

### ISSUE THREAD

**Issue by Ge0rges** [pytorch/pytorch#38581]
## 🐛 Bug

If you set a hook that returns grads of just 0 you expect the weights to not be updated. However using SGD if you set weight decay to a non-zero value, even if such a hook is in place the weight values will still change. 

## To Reproduce
```

import torch.nn as nn
import torch

from torch.utils.data import DataLoader, RandomSampler

# Hook class
class ActiveGradsHook:
    def __init__(self, name):
        self.name = name # Runnable on 1.5

    def __call__(self, grad):
        try:
            return torch.zeros(grad.shape)
        except Exception as e:
            print(e)

def train_new_neurons(model):
    # Generate hooks for each layer
    for name, param in model.named_parameters():
        hook = ActiveGradsHook(name)
        param.register_hook(hook)

    # Train simply
    train(model)


def train(model):
    initial_weights = model.weight.clone()

    optimizer = torch.optim.SGD(model.parameters(), lr=1, momentum=0, weight_decay=1)
    
    inputs = torch.rand(2, 3)
    action_target = torch.rand(2, 2)

    action_output = model(inputs)
    
    loss = nn.MSELoss()(action_target, action_output)

    optimizer.zero_grad()
    loss.backward()
    optimizer.step()

    print(model.weight)
    print(initial_weights)

    assert initial_weights[0][0] == model.weight[0][0] # This fails

if __name__ == "__main__":
    model = nn.Linear(3, 2)
    train_new_neurons(model)

```

## Expected behavior

Gradients are set to 0 so weights shouldn't change. 

## Environment

 - PyTorch Version: 1.5.0
 - OS: Ubuntu 19
 - How you installed PyTorch: pip
 - Python version: 3.8
 - CUDA version: 10.2
 - GPU models and configuration: RTX 2080Ti

**Comment by vballoli** [comment]
IIRC, weight decay isn't related to the gradient, but to the weights(params) itself. Although the gradient is zeroed out, the weight does (and must) get updated since weight decay is applied wrt to the weight. For your expected return, try returning `None` in the hook. See [here](https://pytorch.org/docs/stable/_modules/torch/optim/sgd.html#SGD)

**Comment by agolynski** [comment]
Does zeroing out works if you don't set weight delay?

**Comment by Ge0rges** [comment]
Yes @agolynski

**Comment by agolynski** [comment]
then it seems like WAI? Could you close this if it resolves your concerns?
